File: src/bin/day4.rs
// https://adventofcode.com/2022/day/4

use std::fs::read_to_string;

use itertools::Itertools;

fn parse_input(input: &String) -> Vec<((usize, usize), (usize, usize))> {
    // each line looks like "2-4,6-8"
    input
        .lines()
        .map(|line| {
            let nums = line
                .split(&['-', ','])
                .map(|num_str| num_str.parse::<usize>().unwrap())
                .collect_vec();
            assert_eq!(nums.len(), 4);
            ((nums[0], nums[1]), (nums[2], nums[3]))
        })
        .collect_vec()
}

fn main() {
    let input = read_to_string("input/day4-input.txt").unwrap();
    let range_pairs = parse_input(&input);

    // Count how many ranges are completely contained in their companion range
    let part1_result = range_pairs
        .iter()
        .filter(|((start1, end1), (start2, end2))| {
            (start1 >= start2 && end1 <= end2) || (start2 >= start1 && end2 <= end1)
        })
        .count();

    println!("Part 1 solution = {}", part1_result);

    // Count how many ranges overlap their companion range
    let part2_result = range_pairs
        .iter()
        .filter(|((start1, end1), (start2, end2))| {
            (start1..=end1).contains(&start2) || (start2..=end2).contains(&start1)
        })
        .count();

    println!("Part 2 solution = {}", part2_result);
}

File: src/bin/day4.py
import re
import sys
from typing import List, Tuple


def parse_input(input_file: str) -> List[Tuple[Tuple[int, int], Tuple[int, int]]]:
    rx = re.compile('[,-]')

    with open(input_file) as f:
        range_pairs = []
        for line in f:
            nums = list(map(int, rx.split(line.strip())))
            range_pairs.append(((nums[0], nums[1]), (nums[2], nums[3])))

        return range_pairs


def main():
    range_pairs = parse_input(sys.argv[1])

    part1_solution = 0
    part2_solution = 0
    for (start1, end1), (start2, end2) in range_pairs:
        if (start1 >= start2 and end1 <= end2) or (start2 >= start1 and end2 <= end1):
            part1_solution += 1

        if (start2 <= start1 <= end2) or (start1 <= start2 <= end1):
            part2_solution += 1

    print(f"Part 1 solution = {part1_solution}")
    print(f"Part 2 solution = {part2_solution}")


if __name__ == '__main__':
    main()
